### DISCUSSION TITLE
Deprecate older Python versions?

### DISCUSSION THREAD

**Discussion by JohnVillalovos** [python-gitlab/python-gitlab#1662] [General]
Should we deprecate older Python versions?

We have the 3.0 release coming up. Should we deprecate Python 3.6 support?

Python 3.6.0 was released on 2016-12-23: https://www.python.org/downloads/release/python-360/

One benefit is that Python 3.7 supports: `from __future__ import annotations`

**Comment by JohnVillalovos** [comment]
@nejch What do you think?

**Reply by nejch** [reply]
I'm also eager to start using new feature 😁 

I would maybe keep it at least until its EOL (https://endoflife.date/python, not sure if it's the most accurate prediction but https://www.python.org/dev/peps/pep-0494/#lifespan suggests something like that). Do you mean deprecate and just emit a warning or drop completely? As a library I'd keep it a little conservative, I know Ansible can be like that and the community.general collection depends on this for the gitlab modules.

But maybe that shouldn't stop us from adding `from __future__ import annotations`. We could add https://pypi.org/project/future-annotations/ to requires for python<3.7, it's only a small penalty for 3.6 users at install time.

**Reply by JohnVillalovos** [reply]
I guess by "deprecate" I mean stop supporting. As in drop-completely. Would make sense to me for the 3.0 release. The 2.X release would still have support for Python 3.6.

Looks like Python 3.6 is End-of-Life in less than two months... To me that is pretty much EOL to me.

**Comment by JohnVillalovos** [comment]
Another benefit of moving to Python 3.7 as a minimum is we can start using dataclasses

https://docs.python.org/3/library/dataclasses.html

**Reply by nejch** [reply]
Good point. We could maybe do it in December with 3.0.0.

**Reply by max-wittig** [reply]
Yeah that makes sense. Python 3.6 is officially supported until the 23th of December, according to https://endoflife.date/python

**Reply by JohnVillalovos** [reply]
Cool. So if we have made a decision to drop Python 3.6 support for the 3.0.0 release, we probably could start the work now in the `main` branch.

What do you think?

**Comment by JohnVillalovos** [comment]
I've created a PR to drop support for Python 3.6. And require Python 3.7 or newer.

https://github.com/python-gitlab/python-gitlab/pull/1705
